asserts/snapasserts/snapasserts.go

```diff
@@ -53,42 +55,117 @@ func findSnapDeclaration(snapID, name string, db Finder) (*asserts.SnapDeclarati
+		return "", fmt.Errorf("snap %q file does not have expected size according to signatures (download is broken or tampered): %d != %d", instanceName, snapSize, snapRev.SnapSize())
 	}
 
 	snapID := si.SnapID
 
 	if snapRev.SnapID() != snapID || snapRev.SnapRevision() != si.Revision.N {
-		return fmt.Errorf("snap %q does not have expected ID or revision according to assertions (metadata is broken or tampered): %s / %s != %d / %s", instanceName, si.Revision, snapID, snapRev.SnapRevision(), snapRev.SnapID())
+		return "", fmt.Errorf("snap %q does not have expected ID or revision according to assertions (metadata is broken or tampered): %s / %s != %d / %s", instanceName, si.Revision, snapID, snapRev.SnapRevision(), snapRev.SnapID())
 	}
 
 	snapDecl, err := findSnapDeclaration(snapID, instanceName, db)
 	if err != nil {
-		return err
+		return "", err
 	}
 
 	if snapDecl.SnapName() != snap.InstanceSnap(instanceName) {
-		return fmt.Errorf("cannot install %q, snap %q is undergoing a rename to %q", instanceName, snap.InstanceSnap(instanceName), snapDecl.SnapName())
+		return "", fmt.Errorf("cannot install %q, snap %q is undergoing a rename to %q", instanceName, snap.InstanceSnap(instanceName), snapDecl.SnapName())
+	}
+
+	return CrossCheckProvenance(instanceName, snapRev, snapDecl, model, db)
+}
+
+// CrossCheckProvenance tries to cross check the given snap-revision
+// if it has a non default provenance with the revision-authority
+// constraints of the given snap-declaration including any device
+// scope constraints using model (and implied store).
+// It also returns the provenance if it is different from the default.
+// Ultimately if not default the provenance must also be checked
+// with the provenance in the snap metadata by the caller.
+func CrossCheckProvenance(instanceName string, snapRev *asserts.SnapRevision, snapDecl *asserts.SnapDeclaration, model *asserts.Model, db Finder) (signedProvenance string, err error) {
+	if snapRev.Provenance() == "global-upload" {
+		// nothing to check
+		return "", nil
+	}
+	var store *asserts.Store
+	if model != nil && model.Store() != "" {
+		a, err := db.Find(asserts.StoreType, map[string]string{
+			"store": model.Store(),
+		})
+		if err != nil && !asserts.IsNotFound(err) {
+			return "", err
+		}
+		if a != nil {
+			store = a.(*asserts.Store)
+		}
+	}
+	ras := snapDecl.RevisionAuthority(snapRev.Provenance())
+	matchingRevAuthority := false
+	for _, ra := range ras {
+		if err := ra.Check(snapRev, model, store); err == nil {
+			matchingRevAuthority = true
+			break
+		}
 	}
+	if !matchingRevAuthority {
+		return "", fmt.Errorf("snap %q revision assertion with provenance %q is not signed by an authority authorized on this device: %s", instanceName, snapRev.Provenance(), snapRev.AuthorityID())
+	}
+	return snapRev.Provenance(), nil
+}
 
+// CheckProvenance checks that the given snap has the provided provenance.
```

Comment:
(nitpick^2) The docstring is great, but maybe the function name could already express that this is intended on "trusted" snaps only? like `CheckProvenanceOn{Trusted,Validated,Signed,Safe}Snap()`